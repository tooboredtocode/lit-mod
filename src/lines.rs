use crate::util::slice_like;
use proc_macro::TokenStream;

slice_like!("lines!");

pub fn lines(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as Slice);

    let source = input.source.value();
    let mut result = String::new();

    for range in input.ranges {
        let start = range.0.unwrap_or(0);
        let end = range.1.unwrap_or(input.source.value().len()) - start;

        source.lines().skip(start).take(end).for_each(|s| {
            result.push_str(s);
            result.push('\n');
        });
    }

    if result.ends_with('\n') {
        result.pop();
    }

    LitStr::new(&result, input.source.span())
        .into_token_stream()
        .into()
}

#[cfg(test)]
mod test {
    use trybuild::TestCases;

    #[test]
    fn test_build() {
        let t = TestCases::new();
        t.pass("tests/lines/*.pass.rs");
        t.compile_fail("tests/lines/*.fail.rs");
    }
}
