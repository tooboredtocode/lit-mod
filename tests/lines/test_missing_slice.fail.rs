use lit_mod::lines;

fn main() {
    let s = lines!("Hello, world!");
}