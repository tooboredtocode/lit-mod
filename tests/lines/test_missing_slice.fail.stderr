error: unexpected end of input, lines! expects at least one range
 --> tests/lines/test_missing_slice.fail.rs:4:13
  |
4 |     let s = lines!("Hello, world!");
  |             ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `lines` (in Nightly builds, run with -Z macro-backtrace for more info)
